proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,219 @@
+
+The Type, Import, Module, Instance, and Alias sections can appear in any order
+at the beginning of a module. Each item defined in these sections can only refer
+to previously defined items. For example instance 4 cannot reference instance 5,
+and similarly no instance can refer to the module's own functions.
+
+## Type Section updates
+
+Updates to
+[`typesec`](https://webassembly.github.io/spec/core/binary/modules.html#binary-typesec):
+
+```
+typesec ::=  t*:section_1(vec(type))
+
+type ::=
+            0x60 ft:functype                    ->        ft
+            0x61 mt:moduletype                  ->        mt
+            0x62 it:instancetype                ->        it
+
+functype ::= rt_1:resulttype rt_2:resulttype    ->        rt_1 -> rt-2
+
+moduletype ::=
+  i*:vec(import) e*:vec(exporttype)             ->        {imports i, exports e}
+
+instancetype ::= e*:vec(exporttype)             ->        {exports e}
+
+exporttype ::= nm:name d:importdesc             ->        {name nm, desc d}
+```
+
+referencing:
+
+* [`resulttype`](https://webassembly.github.io/spec/core/binary/types.html#binary-resulttype)
+* [`import`](https://webassembly.github.io/spec/core/binary/modules.html#binary-import)
+
+**Validation**
+
+* each `importdesc` is valid according to import section
+* Types can only reference preceding type definitions. This forces everything to
+  be a DAG and forbids cyclic types.
+
+## Import Section updates
+
+Updates to
+[`importdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-importdesc)
+
+```
+# note that `0x01 0xc0` in MVP-wasm specifies a 1-byte module field of the
+# string 0xc0, but the 0xc0 byte is not valid utf-8, so this was not a valid
+# MVP import
+import ::=
+    ...
+    mod:name  0x01 0xc0 d:importdesc            ->    {module mod, desc d}
+
+importdesc ::=
+    ...
+    0x05 x:typeidx                              ->    module x
+    0x06 x:typeidx                              ->    instance x
+```
+
+The `import` production is modified to allow optionally dropping the `name`
+field. This is done by encoding an invalid utf-8 string so this is an invalid
+MVP import.
+
+**Validation**
+
+* the `func x` production ensures the type `x` is indeed a function type
+* the `module x` production ensures the type `x` is indeed a module type
+* the `instance x` production ensures the type `x` is indeed an instance type
+
+## Module section (100)
+
+A new module section is added
+
+```
+modulesec ::=  t*:section_100(vec(typeidx))           ->        t*
+```
+
+**Validation**
+
+* Each type index `t` points to a module type (e.g. not a function type)
+* This defines the locally defined set of modules, adding to the module index
+  space.
+
+## Instance section (101)
+
+A new section defining local instances
+
+```
+instancesec ::=  i*:section_101(vec(instancedef))     ->    i*
+
+instancedef ::= m:moduleidx e*:vec(exportdesc)        ->    {instantiate m, imports e}
```

Comment:
There's a decent chance that alternatives to `instantiate` could be added in the future (for example, a "tupling" operator that simply creates an instance out of existing fields without needing any module), so could you add some opcode to represent `instantiate`?  `instantiate` will likely appear as a core instruction at some point in the future, so it would be good if this opcode was a valid `instr` opcode

Reply:
👍 added some words, althogh I don't know how best to reserve an opcode so I left it as 0x00 for now with a note saying we'll want to match `instantiate` if it's added as an opcode.

Reply:
Well, probably what will happen is this linking feature would ship before any first-class `instantiate` instruction, so we can just pick one now (probably using the miscellaneous prefix) and prevent anyone future instructions from using it (before `instantiate`)